core/field.ts

```diff
@@ -45,17 +45,55 @@ import * as WidgetDiv from './widgetdiv.js';
 import type {WorkspaceSvg} from './workspace_svg.js';
 import * as Xml from './xml.js';
 
-export type FieldValidator<T = any> = (value?: T) => T|null|undefined;
+/**
+ * Shorthand for `undefined`. Used in cases where a parameter should
+ * always be supplied (so it's not optional), though the value could
+ * be `undefined`.
+ */
+type Un = undefined;
+
+/**
+ * The validation response for an update to a **Field** value.
+ *
+ * @returns `T`, the modified input value to use.
+ * @returns `null` to ignore the input value and invoke `doValueInvalid_`.
+ * @returns `undefined` to use the input value as is for the field update.
+ */
+export type Validation<T> = T|null|undefined;
+
+/**
+ * Where `T` is the value stored on a **Field** and `U` is the value passed into
+ * the constructor and **setValue**, `T` and `U` are the same for most
+ * subclasses. If they are different and `U` is not compatible with `T`, it is
+ * invalid to return `undefined` since the input value *must be transformed*
+ * to store as a `T`. For example, a subclass could take in an image then use
+ * **doClassValidation_** to convert it to a string.
+ */
+type ClassValidation<T, U> = U extends T ? Validation<T>: T|null;
```

Comment:
I got rid of `ClassValidation` in my update to `doClassValidation_`, which now enforces this rule in addition to enforcing subclasses not to return `undefined` if `newValue === undefined`. My hope is that the new implementation of `doClassValidation_` is clearer than this type.

Regarding keeping or removing `Validation`, I think the important thing is just that users have in scope the context that `undefined` and `null` are meaningful. I can just move the corresponding comments to `FieldValidator` from `Validation`.

I think `Validation` was more important early on when there wasn't a `U` since, in the context of `setValue`, there wasn't a distinction between the initially supplied value and the processed value. The implementation was very confusing to me early on... it read something like, *"This is `newValue`, but it's sometimes (not always!) an `Error`,  but other times it's the `newValue` with some small changes to it, but other times its actually something completely different, oh and `newValue === undefined` actually means the class validation is `null`, but if class validation is `undefined` that means `newValue !== undefined`..."*  😆 It lead to a lot of mental juggling of what-was-what between `setValue`, `doClassValidation_` and `processValidation_`, and `FieldValidator`. The `Validation` type was a way to split what `T`, `undefined`, and `null` meant in the various contexts.